chiron/multistate.py

```diff
@@ -0,0 +1,701 @@
+from chiron.neighbors import NeighborListNsqrd
+from openmm import unit
+import numpy as np
+from chiron.mcmc import MCMCMove, MCMCSampler
+from chiron.reporters import MultistateReporter
+
+
+class MultiStateSampler:
+    """
+    A sampler for simulating multiple thermodynamic states using replicas.
+
+    This class provides a general simulation facility for sampling from multiple
+    thermodynamic states. It allows specifying any set of thermodynamic states.
+    If instantiated on its own, the thermodynamic state indices associated with
+    each state are specified, and replica mixing does not change any thermodynamic states,
+    meaning that each replica remains in its original thermodynamic state.
+
+    Attributes
+    ----------
+    n_states : int
+        Number of thermodynamic states (read-only).
+    n_replicas : int
+        Number of replicas (read-only).
+    iteration : int
+        Current iteration of the simulation (read-only).
+    mcmc_sampler : MCMCSampler
+        MCMC sampler used to propagate the simulation.
+    sampler_states : List[SamplerState]
+        Sampler states list at the current iteration.
+    is_periodic : bool
+        True if system is periodic, False if not, None if not initialized.
+    is_completed : bool
+        Check if the sampler has reached any stop target criteria (read-only).
+
+    Methods
+    -------
+    create(thermodynamic_states: List[ThermodynamicState], sampler_states: List[SamplerState], nbr_list: NeighborListNsqrd)
+        Creates a new multistate sampler simulation.
+    minimize(tolerance: unit.Quantity = 1.0 * unit.kilojoules_per_mole / unit.nanometers, max_iterations: int = 1000)
+        Minimizes all replicas in the sampler.
+    run(n_iterations: int = 10)
+        Executes the replica-exchange simulation for a specified number of iterations.
+
+    """
+
+    def __init__(
+        self,
+        mcmc_sampler: MCMCSampler,
+        reporter: MultistateReporter,
+    ):
+        """
+        Initialize the MultiStateSampler.
+
+        Parameters
+        ----------
+        mcmc_sampler : MCMCSampler
+            The MCMCSampler used to propagate the thermodynamic states.
+        reporter : MultistateReporter
+            The reporter used to store the simulation data.
+        """
+
+        import copy
+        from chiron.analysis import MBAREstimator
+
+        # These will be set on initialization. See function
+        # create() for explanation of single variables.
+        self._thermodynamic_states = None
+        self._unsampled_states = None
+        self._sampler_states = None
+        self._replica_thermodynamic_states = None
+        self._iteration = None
+        self._energy_thermodynamic_states = None
+        self._neighborhoods = None
+        self._n_accepted_matrix = None
+        self._n_proposed_matrix = None
+
+        self._reporter = reporter  # NOTE: reporter needs to be putlic, API change ahead
```

Comment:
minor typo in the comment. 
```suggestion
        self._reporter = reporter  # NOTE: reporter needs to be public, API change ahead
```